subsys/usb/usbc/usbc_stack.c

```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright (c) 2022 The Chromium OS Authors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT usbc_port
+
+#include <zephyr/devicetree.h>
+#include <zephyr/init.h>
+#include <zephyr/smf.h>
+#include <zephyr/usbc/usbc.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+#include "usbc_stack.h"
+
+LOG_MODULE_REGISTER(usbc_stack);
+
+static const struct usbc_subsystem_api usbc_api;
+static int usbc_subsys_init(const struct device *dev);
+
+#define THREAD_STACK(inst) K_THREAD_STACK_DEFINE(my_stack_area_##inst, CONFIG_USBC_STACK_SIZE)
+
+#define PORT_THREAD(inst) void run_usbc_##inst(void *port_dev, void *unused1, void *unused2) \
+	{										     \
+		const struct device *dev = (const struct device *)port_dev;		     \
+		struct usbc_port_data *port = dev->data;				     \
+		struct request_value *req;						     \
+		int32_t request;							     \
+											     \
+		while (1) {								     \
+			req = k_fifo_get(&port->request_fifo, K_NO_WAIT);		     \
+			request = (req != NULL) ? req->val : REQUEST_NOP;		     \
+			pe_run(dev, request);					     \
+			prl_run(dev);						     \
+			tc_run(dev, request);					     \
+			if (request == PRIV_PORT_REQUEST_SUSPEND) {			     \
+				k_thread_suspend(port->port_thread);			     \
+			}								     \
+			k_msleep(CONFIG_USBC_STATE_MACHINE_CYCLE_TIME);			     \
+		}									     \
```

Comment:
I initially thought about using a workqueue but decided against it because the USBC timing requirements might not be met if the system is loaded.

As for tolerance, I've only tested the code running with one USBC port.